File: app/src/main/java/com/manimarank/spell4wiki/views/EndlessRecyclerView.java
package com.manimarank.spell4wiki.views;

import android.content.Context;
import android.util.AttributeSet;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.manimarank.spell4wiki.adapters.EndlessRecyclerAdapter;
import com.manimarank.spell4wiki.utils.NetworkUtils;

import java.util.List;

public class EndlessRecyclerView extends RecyclerView {


    private boolean isLoading = false;
    private EndlessListener listener;
    private EndlessRecyclerAdapter adapter;
    private boolean isLastPage = false;
    private boolean loadMoreEnabled = true;


    public EndlessRecyclerView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public EndlessRecyclerView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public EndlessRecyclerView(Context context) {
        super(context);
    }

    public void setAdapter(EndlessRecyclerAdapter mAdapter, @NonNull LinearLayoutManager layoutManager) {
        super.setAdapter(mAdapter);
        this.adapter = mAdapter;
        removeLoader();

        this.addOnScrollListener(new OnScrollListener() {

            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                int visibleItemCount = layoutManager.getChildCount();
                int totalItemCount = layoutManager.getItemCount();
                int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();

                if (!isLoading && loadMoreEnabled) {
                    if ((visibleItemCount + firstVisibleItemPosition) >= totalItemCount && firstVisibleItemPosition >= 0 && listener != null) {
                        if (isLastPage || !NetworkUtils.INSTANCE.isConnected(getContext()))
                            listener.loadFail();
                        else if (listener.loadData())
                            addLoaded();
                    }
                }
            }
        });
    }

    public void addNewData(List<String> data) {
        removeLoader();
        this.adapter.addItems(data);
    }

    public boolean isLastPage() {
        return isLastPage;
    }

    public void removeLoader() {
        isLoading = false;
        this.adapter.removeLoading();
    }

    private void addLoaded() {
        isLoading = true;
        this.adapter.addLoading();
    }


    public void reset() {
        this.adapter.clear();
        addLoaded();
        isLastPage = false;
        enableLoadMore();
        super.setAdapter(adapter);
    }

    public void setListener(EndlessListener listener) {
        this.listener = listener;
    }

    public void setLastPage() {
        this.isLastPage = true;
    }

    public void disableLoadMore() {
        loadMoreEnabled = false;
    }

    public void enableLoadMore() {
        loadMoreEnabled = true;
    }


    public interface EndlessListener {
        boolean loadData();

        void loadFail();
    }

}

File: app/src/main/java/com/manimarank/spell4wiki/ui/custom/EndlessRecyclerView.kt
package com.manimarank.spell4wiki.ui.custom

import android.content.Context
import android.util.AttributeSet
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.manimarank.spell4wiki.ui.spell4wiktionary.EndlessRecyclerAdapter
import com.manimarank.spell4wiki.utils.NetworkUtils.isConnected

/**
 * Endless Recycler View Widget component for lazy loading
 * @property isLoading Boolean
 * @property listener EndlessListener?
 * @property adapter EndlessRecyclerAdapter?
 * @property isLastPage Boolean
 * @property loadMoreEnabled Boolean
 */
class EndlessRecyclerView : RecyclerView {
    private var isLoading = false
    private var listener: EndlessListener? = null
    private var adapter: EndlessRecyclerAdapter? = null
    var isLastPage = false
        private set
    private var loadMoreEnabled = true

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle)
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    constructor(context: Context) : super(context)

    fun setAdapter(mAdapter: EndlessRecyclerAdapter?, layoutManager: LinearLayoutManager) {
        super.setAdapter(mAdapter)
        adapter = mAdapter
        removeLoader()
        addOnScrollListener(object : OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                val visibleItemCount = layoutManager.childCount
                val totalItemCount = layoutManager.itemCount
                val firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition()
                if (!isLoading && loadMoreEnabled) {
                    if (visibleItemCount + firstVisibleItemPosition >= totalItemCount && firstVisibleItemPosition >= 0 && listener != null) {
                        if (isLastPage || !isConnected(context))
                            listener?.loadFail()
                        else if (listener?.loadData() == true)
                            addLoaded()
                    }
                }
            }
        })
    }

    fun addNewData(data: MutableList<String>?) {
        removeLoader()
        if (data != null)
            adapter?.addItems(data)
    }

    fun removeLoader() {
        isLoading = false
        adapter?.removeLoading()
    }

    private fun addLoaded() {
        isLoading = true
        adapter?.addLoading()
    }

    fun reset() {
        adapter?.clear()
        addLoaded()
        isLastPage = false
        enableLoadMore()
        super.setAdapter(adapter)
    }

    fun setListener(listener: EndlessListener?) {
        this.listener = listener
    }

    fun setLastPage() {
        isLastPage = true
    }

    fun disableLoadMore() {
        loadMoreEnabled = false
    }

    fun enableLoadMore() {
        loadMoreEnabled = true
    }

    interface EndlessListener {
        fun loadData(): Boolean
        fun loadFail()
    }
}